src/lib/core/trigger.cc

```diff
@@ -54,24 +56,124 @@ trigger_fiber_f(va_list ap)
 	return trigger->run(trigger, event);
 }
 
+/**
+ * A single link in a list of triggers scheduled for execution. Can't be part of
+ * struct trigger, because one trigger might be part of an unlimited number of
+ * run lists.
+ */
+struct run_link {
+	/** A link in the run list belonging to one trigger_run() call.  */
+	struct rlist in_run;
+	/**
+	 * A link in a list of all links referencing that trigger. Used to
+	 * remove the trigger from whatever list it's on during trigger_clear().
+	 */
+	struct rlist in_trigger;
+	/** The trigger to be executed. */
+	struct trigger *trigger;
+};
+
+/** A memory pool for trigger run link allocation. */
+static __thread struct mempool run_link_pool;
+
+/** Add the trigger to the run list. */
+static int
+run_list_put_trigger(struct rlist *list, struct trigger *trigger)
+{
+	struct run_link *run_link =
+		(struct run_link *)mempool_alloc(&run_link_pool);
+	if (run_link == NULL) {
+		diag_set(OutOfMemory, sizeof(struct run_link),
+			 "mempool_alloc", "trigger run link");
+		return -1;
+	}
+	run_link->trigger = trigger;
+	rlist_add_tail_entry(list, run_link, in_run);
+	rlist_add_tail_entry(&trigger->run_links, run_link, in_trigger);
+	return 0;
+}
+
+/** Take the next trigger from the run list and free the corresponding link. */
+static inline struct trigger *
+run_list_take_trigger(struct rlist *list)
+{
+	struct run_link *link = rlist_shift_entry(list, struct run_link,
+						  in_run);
+	struct trigger *trigger = link->trigger;
+	rlist_del_entry(link, in_trigger);
+	mempool_free(&run_link_pool, link);
+	return trigger;
+}
+
+/** Execute the triggers in an order specified by \a list. */
+static inline int
+trigger_run_list(struct rlist *list, void *event)
+{
+	while (!rlist_empty(list)) {
+		struct trigger *trigger = run_list_take_trigger(list);
+		if (trigger->run(trigger, event) != 0) {
+			while (!rlist_empty(list))
+				run_list_take_trigger(list);
```

Comment:
This loop is used in three places. Would be nice to put it in a helper function (`run_list_clear`?)

Reply:
Sure:
<details><summary>diff</summary>

```diff
diff --git a/src/lib/core/trigger.cc b/src/lib/core/trigger.cc
index 9a53c9284..a70f4b54d 100644
--- a/src/lib/core/trigger.cc
+++ b/src/lib/core/trigger.cc
@@ -102,6 +102,14 @@ run_list_take_trigger(struct rlist *list)
        return trigger;
 }
 
+/** Empty the run list and free all the links allocated for it. */
+static inline void
+run_list_clear(struct rlist *list)
+{
+       while (!rlist_empty(list))
+               run_list_take_trigger(list);
+}
+
 /** Execute the triggers in an order specified by \a list. */
 static inline int
 trigger_run_list(struct rlist *list, void *event)
@@ -109,8 +117,7 @@ trigger_run_list(struct rlist *list, void *event)
        while (!rlist_empty(list)) {
                struct trigger *trigger = run_list_take_trigger(list);
                if (trigger->run(trigger, event) != 0) {
-                       while (!rlist_empty(list))
-                               run_list_take_trigger(list);
+                       run_list_clear(list);
                        return -1;
                }
        }
@@ -138,8 +145,7 @@ trigger_run(struct rlist *list, void *event)
        struct trigger *trigger;
        rlist_foreach_entry(trigger, list, link) {
                if (run_list_put_trigger(&run_list, trigger) != 0) {
-                       while (!rlist_empty(&run_list))
-                               run_list_take_trigger(&run_list);
+                       run_list_clear(list);
                        return -1;
                }
        }
@@ -153,8 +159,7 @@ trigger_run_reverse(struct rlist *list, void *event)
        struct trigger *trigger;
        rlist_foreach_entry_reverse(trigger, list, link) {
                if (run_list_put_trigger(&run_list, trigger) != 0) {
-                       while (!rlist_empty(&run_list))
-                               run_list_take_trigger(&run_list);
+                       run_list_clear(list);
                        return -1;
                }
        }

```

</diff>